embassy-stm32/src/dac/v2.rs

```diff
@@ -0,0 +1,295 @@
+            Ch1Trigger::Exti9 => dac_v2::vals::Tsel1::EXTI9,
+            Ch1Trigger::Software => dac_v2::vals::Tsel1::SOFTWARE,
+        }
+    }
+}
+
+pub enum Ch2Trigger {
+    Tim6,
+    Tim8,
+    Tim7,
+    Tim5,
+    Tim2,
+    Tim4,
+    Exti9,
+    Software,
+}
+
+impl Ch2Trigger {
+    fn tsel(&self) -> dac_v2::vals::Tsel2 {
+        match self {
+            Ch2Trigger::Tim6 => dac_v2::vals::Tsel2::TIM6_TRGO,
+            Ch2Trigger::Tim8 => dac_v2::vals::Tsel2::TIM8_TRGO,
+            Ch2Trigger::Tim7 => dac_v2::vals::Tsel2::TIM7_TRGO,
+            Ch2Trigger::Tim5 => dac_v2::vals::Tsel2::TIM5_TRGO,
+            Ch2Trigger::Tim2 => dac_v2::vals::Tsel2::TIM2_TRGO,
+            Ch2Trigger::Tim4 => dac_v2::vals::Tsel2::TIM4_TRGO,
+            Ch2Trigger::Exti9 => dac_v2::vals::Tsel2::EXTI9,
+            Ch2Trigger::Software => dac_v2::vals::Tsel2::SOFTWARE,
+        }
+    }
+}
+
+pub enum Alignment {
+    Left,
+    Right,
+}
+
+pub enum Value {
+    Bit8(u8),
+    Bit12(u16, Alignment),
+}
+
+pub struct Dac<'d, T: Instance> {
+    //peri: T,
+    ch1: Option<AnyPin>,
+    ch2: Option<AnyPin>,
+    phantom: PhantomData<&'d mut T>,
+}
+
+impl<'d, T: Instance> Dac<'d, T> {
+    pub fn new(
+        peri: impl Unborrow<Target = T> + 'd,
+        ch1: impl Unborrow<Target = impl DacPin<T, 1>>,
+        ch2: impl Unborrow<Target = impl DacPin<T, 2>>,
+    ) -> Self {
+        unborrow!(peri);
+        unborrow!(ch1, ch2);
+
+        let ch1 = Some(ch1.degrade());
+        unsafe {
+            T::regs().cr().modify(|reg| {
+                reg.set_en1(true);
+            });
+        }
+
+        let ch2 = Some(ch2.degrade());
+        unsafe {
+            T::regs().cr().modify(|reg| {
+                reg.set_en2(true);
+            });
+        }
+
+        let mut dac = Self {
+            ch1,
+            ch2,
+            phantom: PhantomData,
+        };
+
+        dac
+    }
+
+    pub fn new_ch1(
+        peri: impl Unborrow<Target = T> + 'd,
+        ch1: impl Unborrow<Target = impl DacPin<T, 1>>,
+    ) -> Self {
+        unborrow!(peri);
+        unborrow!(ch1);
+
+        let ch1 = Some(ch1.degrade());
+        unsafe {
+            T::regs().cr().modify(|reg| {
+                reg.set_en1(true);
+            });
+        }
+
+        let mut dac = Self {
+            ch1,
+            ch2: None,
+            phantom: PhantomData,
+        };
+
+        dac
+    }
+
+    pub fn new_ch2(
+        peri: impl Unborrow<Target = T> + 'd,
+        ch2: impl Unborrow<Target = impl DacPin<T, 2>>,
+    ) -> Self {
+        unborrow!(peri);
+        unborrow!(ch2);
+
+        let ch2 = Some(ch2.degrade());
+        unsafe {
+            T::regs().cr().modify(|reg| {
+                reg.set_en2(true);
+            });
+        }
+
+        let mut dac = Self {
+            ch1: None,
+            ch2,
+            phantom: PhantomData,
+        };
+
+        dac
+    }
+
+    pub fn enable_ch1(&mut self) -> Result<(), Error> {
```

Comment:
There's lots of functions that are duplicated for ch1 and ch2. It's a much nicer API (and less code duplication) if you make them `fn enable(&mut self, ch: usize)`.

Applies to enable, disable, select_trigger, trigger, set.

You'll have to arrayify the regs in stm32-data probably.